src/tools/rbd_mirror/GroupReplayer.cc

```diff
@@ -0,0 +1,1305 @@
+        snap_spec.image_id = image.spec.image_id;
+        snap_spec.snap_id = snap_info.id;
+
+        local_image_snap_specs.push_back(snap_spec);
+      }
+    }
+  }
+
+  group_snap.snaps = local_image_snap_specs;
+  group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  librbd::cls_client::group_snap_set(&op, group_snap);
+
+  auto comp = create_rados_callback(
+      new LambdaContext([this](int r) {
+        handle_create_regular_group_snapshot(r);
+      }));
+  int r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_ctx.group_id), comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void GroupReplayer<I>::handle_create_regular_group_snapshot(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    derr << "error creating local non-primary group snapshot: "
+         << cpp_strerror(r) << dendl;
+  }
+
+  return;
+}
+
+template <typename I>
+int GroupReplayer<I>::local_group_image_list_by_id(
+    std::vector<cls::rbd::GroupImageStatus> *image_ids) {
+  std::string group_header_oid = librbd::util::group_header_name(
+      m_local_group_ctx.group_id);
+
+  dout(10) << "listing images in local group id " << group_header_oid << dendl;
+  image_ids->clear();
+
+  int r = 0;
+  const int max_read = 1024;
+  cls::rbd::GroupImageSpec start_last;
+  do {
+    std::vector<cls::rbd::GroupImageStatus> image_ids_page;
+
+    r = librbd::cls_client::group_image_list(&m_local_io_ctx, group_header_oid,
+                                             start_last, max_read,
+                                             &image_ids_page);
+
+    if (r < 0) {
+      derr << "error reading image list from local group: "
+           << cpp_strerror(-r) << dendl;
+      return r;
+    }
+    image_ids->insert(image_ids->end(), image_ids_page.begin(),
+                      image_ids_page.end());
+
+    if (image_ids_page.size() > 0)
+      start_last = image_ids_page.rbegin()->spec;
+
+    r = image_ids_page.size();
+  } while (r == max_read);
+
+  return 0;
+}
+
+template <typename I>
+void GroupReplayer<I>::create_mirror_snapshot_start(
+    const cls::rbd::MirrorSnapshotNamespace &remote_group_snap_ns,
+    ImageReplayer<I> *image_replayer, int64_t *local_group_pool_id,
+    std::string *local_group_id, std::string *local_group_snap_id,
+    Context *on_finish) {
+  dout(20) << remote_group_snap_ns << " " << image_replayer << dendl;
+
+  ceph_assert(remote_group_snap_ns.is_primary());
+
+  std::unique_lock locker{m_lock};
+
+  std::vector<cls::rbd::GroupImageStatus> local_images;
+  for (auto it = m_remote_group_snaps.begin(); it != m_remote_group_snaps.end(); ++it) {
+    auto snap_type = cls::rbd::get_group_snap_namespace_type(
+        it->second.snapshot_namespace);
+    if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_USER) {
+      if (local_images.empty()) {
+        int r = local_group_image_list_by_id(&local_images);
```

Comment:
Is it possible that the list images that were part of the group when this snap was created is not the same as when this call is executed?